internal/otelarrow/admission2/boundedqueue.go

```diff
@@ -0,0 +1,177 @@
+// - element=non-nil, error=non-nil: the slow success path
+func (bq *BoundedQueue) acquireOrGetWaiter(pending uint64) (*list.Element, error) {
+	bq.lock.Lock()
+	defer bq.lock.Unlock()
+
+	if pending > bq.maxLimitAdmit {
+		// when the request will never succeed because it is
+		// individually over the total limit, fail fast.
+		return nil, ErrRequestTooLarge
+	}
+
+	if bq.currentAdmitted+pending <= bq.maxLimitAdmit {
+		// the fast success path.
+		bq.currentAdmitted += pending
+		return nil, nil
+	}
+
+	// since we were unable to admit, check if we can wait.
+	if bq.currentWaiting+pending > bq.maxLimitWait {
+		return nil, ErrTooMuchWaiting
+	}
+
+	// otherwise we need to wait
+	return bq.addWaiterLocked(pending), nil
+}
+
+// Acquire implements Queue.
+func (bq *BoundedQueue) Acquire(ctx context.Context, pending uint64) (ReleaseFunc, error) {
+	element, err := bq.acquireOrGetWaiter(pending)
+	parentSpan := trace.SpanFromContext(ctx)
+	pendingAttr := trace.WithAttributes(attribute.Int64("pending", int64(pending)))
+
+	if err != nil {
+		parentSpan.AddEvent("admission rejected (fast path)", pendingAttr)
+		return noopRelease, err
+	} else if element == nil {
+		parentSpan.AddEvent("admission accepted (fast path)", pendingAttr)
+		return bq.releaseFunc(pending), nil
+	}
+
+	parentSpan.AddEvent("enter admission queue")
+
+	ctx, span := bq.tracer.Start(ctx, "admission_blocked", pendingAttr)
+	defer span.End()
+
+	waiter := element.Value.(*waiter)
+
+	select {
+	case <-waiter.notify.Chan():
+		parentSpan.AddEvent("admission accepted (slow path)", pendingAttr)
+		return bq.releaseFunc(pending), nil
+
+	case <-ctx.Done():
+		bq.lock.Lock()
+		defer bq.lock.Unlock()
+
+		if waiter.notify.HasBeenNotified() {
+			// We were also admitted, which can happen
+			// concurrently with cancellation. Make sure
+			// to release since no one else will do it.
+			bq.releaseLocked(pending)
+		} else {
+			// Remove ourselves from the list of waiters
+			// so that we can't be admitted in the future.
+			bq.removeWaiterLocked(pending, element)
+			bq.admitWaitersLocked()
+		}
+
+		parentSpan.AddEvent("admission rejected (canceled)", pendingAttr)
+		return noopRelease, status.Error(grpccodes.Canceled, context.Cause(ctx).Error())
+	}
+}
+
+func (bq *BoundedQueue) admitWaitersLocked() {
+	for bq.waiters.Len() != 0 {
+		// Ensure there is enough room to admit the next waiter.
+		element := bq.waiters.Back()
+		waiter := element.Value.(*waiter)
+		if bq.currentAdmitted+waiter.pending > bq.maxLimitAdmit {
+			// Returning means continuing to wait for the
+			// most recent arrival to get service by another release.
+			return
+		}
+
+		// Release the next waiter and tell it that it has been admitted.
+		bq.removeWaiterLocked(waiter.pending, element)
+		bq.currentAdmitted += waiter.pending
+
+		waiter.notify.Notify()
+	}
+}
+
+func (bq *BoundedQueue) addWaiterLocked(pending uint64) *list.Element {
+	bq.currentWaiting += pending
+	return bq.waiters.PushBack(&waiter{
```

Comment:
Using PushBack() and Back() gives us LIFO because the item that was last-inserted (via PushBack) is the item that is first removed (via Back).

The use of PushBack() and Front() would give us FIFO.

Reply:
Ahh thanks my thinking was backwards, it is indeed LIFO